File: client2.c
/*
 * Auth: ASyS
 *
 * File: client2.c
 *
 * Desc: Modified client program "client.c" that takes SERVER_IP_ADDRESS and
 *       SERVER_PORT_NUMBER as argument in addition to SERVER_INDEX and attempt
 *       to connect to them. Upon successful connection, it reads the first 
 *	 1024 characters sent to it and puts it to the STDOUT.
 *
 * Err : Error handling is minimal or non-existent for most part except 
 *       when socket creation failed or connection to server failed in which
 *       error message will be displayed in addition to non-zero return
 *       value. SERVER_IP_ADDRESS, SERVER_PORT_ADDRESS and SERVER_INDEX
 *       validity is not verified. Therefore, wrong formatting behavior is
 *       not known...(At least to the author ;-).
 */
#include <stdlib.h>//atoi() ascii to integer
#include <stdio.h>
#include <sys/socket.h>
#include <sys/types.h>
#include <arpa/inet.h>//htons()
#include <netinet/in.h>//struct sockaddr_in
#include <string.h>//strcat, strcpy

int main(int argc, char **argv)
{
    int sockd;//Socket descriptor for this client

    if((sockd = socket(AF_INET, SOCK_STREAM, 0)) ==1) {
	fprintf(stderr, "Socket creation failed\n");
	exit(1);
    }
    struct sockaddr_in svr;//Struct used to store target server address
    svr.sin_addr.s_addr = inet_addr(argv[1]);//Destination server IP add.
					     //+assigned from cmdline arg.
    svr.sin_family = AF_INET;
    svr.sin_port = htons(atoi(argv[2]));//Destination port
    
    printf("Connecting to %s:%s\n", argv[1], argv[2]);
    if(connect(sockd, (struct sockaddr *)&svr, sizeof svr) < 0){
	fprintf(stderr, "Connection error\n");
	exit(2);
    }
    char svri[10];//Hold server index
    strcpy(svri,argv[3]);//Copy user supplied index number
    strcat(svri, "\n");//Append a newline to server index
    send(sockd, svri, sizeof svri, 0);//Send server index to load balancer
    char msg[1024];
    recv(sockd, msg, sizeof msg, 0);
    printf(msg);
}

File: Client2.java
/*
 * Auth: ASyS
 *
 * File: Client2.java
 *
 * Desc: Client program that takes SERVER_IP_ADDRESS, SERVER_PORT_NUMBER and
 *       SERVER_INDEX as arguments and attempt to connect to the load
 *       balancer. Once connected, the client sends the SERVER_INDEX to the
 *       load balancer, reads a single line from the load balancer and print
 *       it on the STDOUT. 
 *
 * Err : Malformed SERVER_IP_ADDRESS causes unknown host exception to
 *       be thrown and the client to exit with code 1. Any error in IO throws
 *       IO exception with exit code 2.
 */
import java.net.*;
import java.io.*;

class Client2 {
    public static void main(String... args) {
	System.out.println(
	    "Connecting to " + args[0] + ":" + args[1] + "...");
    try {
	Socket socket = new Socket(args[0], Integer.parseInt(args[1]));

	new PrintWriter(
	    socket.getOutputStream(), true
	).println(args[2]);

	System.out.println(
	    new BufferedReader(
		new InputStreamReader(
		    socket.getInputStream()
		)
	    ).readLine()
	);
	socket.close();
    } catch(UnknownHostException uhe) {
	System.err.println("Unknown host: " + args[0]);
	System.exit(1);
    } catch(IOException ioe) {
	System.err.println("I/O connection failed");
	System.exit(2);
    }
    }
}
